File: google-sheets/sheets_client.py
"""
Cliente para ler e escrever planilhas no Google Sheets via API.

Requisitos:
1. Projeto no Google Cloud com Google Sheets API habilitada
2. Conta de serviço + arquivo JSON de credenciais
3. Planilha compartilhada com o e-mail da conta de serviço (Editor)
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Any

import gspread
from dotenv import load_dotenv
from google.oauth2.service_account import Credentials

SCOPES = [
    "https://www.googleapis.com/auth/spreadsheets",
    "https://www.googleapis.com/auth/drive.readonly",
]

BASE_DIR = Path(__file__).resolve().parent


def _load_env() -> None:
    load_dotenv(BASE_DIR / ".env")


def get_credentials(credentials_file: str | None = None) -> Credentials:
    _load_env()
    path = credentials_file or os.getenv("GOOGLE_CREDENTIALS_FILE", "credentials.json")
    credentials_path = Path(path)
    if not credentials_path.is_absolute():
        credentials_path = BASE_DIR / credentials_path

    if not credentials_path.exists():
        raise FileNotFoundError(
            f"Arquivo de credenciais não encontrado: {credentials_path}\n"
            "Baixe o JSON da conta de serviço no Google Cloud Console "
            "e salve como credentials.json nesta pasta."
        )

    return Credentials.from_service_account_file(str(credentials_path), scopes=SCOPES)


def get_client(credentials_file: str | None = None) -> gspread.Client:
    credentials = get_credentials(credentials_file)
    return gspread.authorize(credentials)


def worksheet_by_gid(spreadsheet: gspread.Spreadsheet, gid: int | str) -> gspread.Worksheet:
    gid_int = int(gid)
    for worksheet in spreadsheet.worksheets():
        if worksheet.id == gid_int:
            return worksheet
    raise ValueError(f"Aba com gid={gid_int} não encontrada na planilha '{spreadsheet.title}'")


def open_spreadsheet(
    spreadsheet_id: str | None = None,
    credentials_file: str | None = None,
) -> gspread.Spreadsheet:
    _load_env()
    sheet_id = spreadsheet_id or os.getenv("SPREADSHEET_ID")
    if not sheet_id:
        raise ValueError(
            "Informe spreadsheet_id ou defina SPREADSHEET_ID no arquivo .env"
        )

    client = get_client(credentials_file)
    return client.open_by_key(sheet_id)


def get_worksheet(
    sheet_name: str | None = None,
    spreadsheet_id: str | None = None,
    sheet_gid: int | str | None = None,
    credentials_file: str | None = None,
) -> gspread.Worksheet:
    _load_env()
    name = sheet_name or os.getenv("SHEET_NAME") or None
    spreadsheet = open_spreadsheet(spreadsheet_id, credentials_file)

    if sheet_gid is not None:
        return worksheet_by_gid(spreadsheet, sheet_gid)
    if name:
        return spreadsheet.worksheet(name)
    return spreadsheet.sheet1


def read_all(
    sheet_name: str | None = None,
    spreadsheet_id: str | None = None,
) -> list[list[Any]]:
    worksheet = get_worksheet(sheet_name, spreadsheet_id)
    return worksheet.get_all_values()


def read_range(
    range_notation: str,
    sheet_name: str | None = None,
    spreadsheet_id: str | None = None,
) -> list[list[Any]]:
    worksheet = get_worksheet(sheet_name, spreadsheet_id)
    return worksheet.get(range_notation)


def write_cell(
    cell: str,
    value: Any,
    sheet_name: str | None = None,
    spreadsheet_id: str | None = None,
) -> None:
    worksheet = get_worksheet(sheet_name, spreadsheet_id)
    worksheet.update_acell(cell, value)


def write_range(
    range_notation: str,
    values: list[list[Any]],
    sheet_name: str | None = None,
    spreadsheet_id: str | None = None,
) -> None:
    worksheet = get_worksheet(sheet_name, spreadsheet_id)
    worksheet.update(range_notation, values)


def append_rows(
    rows: list[list[Any]],
    sheet_name: str | None = None,
    spreadsheet_id: str | None = None,
) -> None:
    worksheet = get_worksheet(sheet_name, spreadsheet_id)
    worksheet.append_rows(rows, value_input_option="USER_ENTERED")

File: google-sheets/sheets-client.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import dotenv from "dotenv";
import { google } from "googleapis";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: path.join(__dirname, ".env") });

const SCOPES = [
  "https://www.googleapis.com/auth/spreadsheets",
  "https://www.googleapis.com/auth/drive.readonly",
];

export function getAuth() {
  const credsFile = process.env.GOOGLE_CREDENTIALS_FILE || "credentials.json";
  const credsPath = path.join(__dirname, credsFile);

  if (!fs.existsSync(credsPath)) {
    throw new Error(`Credenciais não encontradas: ${credsPath}`);
  }

  return new google.auth.GoogleAuth({
    keyFile: credsPath,
    scopes: SCOPES,
  });
}

export async function getSheetsClient() {
  const auth = getAuth();
  return google.sheets({ version: "v4", auth });
}

export async function getSpreadsheetMeta(spreadsheetId) {
  const sheets = await getSheetsClient();
  const { data } = await sheets.spreadsheets.get({ spreadsheetId });
  return data;
}

export function getWorksheetByGid(spreadsheet, gid) {
  const gidNum = Number(gid);
  const sheet = spreadsheet.sheets?.find((s) => s.properties?.sheetId === gidNum);
  if (!sheet?.properties?.title) {
    throw new Error(`Aba com gid=${gidNum} não encontrada`);
  }
  return sheet.properties;
}

export async function readAll(spreadsheetId, sheetName) {
  const sheets = await getSheetsClient();
  const range = sheetName ? `'${sheetName.replace(/'/g, "''")}'` : undefined;
  const { data } = await sheets.spreadsheets.values.get({
    spreadsheetId,
    range: range || "A:ZZ",
  });
  return data.values || [];
}

export async function writeCell(spreadsheetId, sheetName, cell, value) {
  const sheets = await getSheetsClient();
  const range = `'${sheetName.replace(/'/g, "''")}'!${cell}`;
  await sheets.spreadsheets.values.update({
    spreadsheetId,
    range,
    valueInputOption: "USER_ENTERED",
    requestBody: { values: [[value]] },
  });
}

export async function appendRows(spreadsheetId, sheetName, rows) {
  const sheets = await getSheetsClient();
  const range = `'${sheetName.replace(/'/g, "''")}'!A:ZZ`;
  await sheets.spreadsheets.values.append({
    spreadsheetId,
    range,
    valueInputOption: "USER_ENTERED",
    insertDataOption: "INSERT_ROWS",
    requestBody: { values: rows },
  });
}

export const SHEETS_CONFIG = [
  { label: "Planilha 1", idEnv: "SPREADSHEET_ID_1", gidEnv: "SPREADSHEET_GID_1" },
  { label: "Planilha 2", idEnv: "SPREADSHEET_ID_2", gidEnv: "SPREADSHEET_GID_2" },
];
